python/pyarrow/types.pxi

```diff
@@ -614,6 +614,76 @@ cdef class UnknownExtensionType(ExtensionType):
         return self.serialized
 
 
+cdef class GenericExtensionType(BaseExtensionType):
+    """
+    Concrete base class for Python-defined extension types.
+    """
+
+    def __cinit__(self):
+        if type(self) is GenericExtensionType:
+            raise TypeError("Can only instantiate subclasses of "
+                            "GenericExtensionType")
+
+    def __init__(self, DataType storage_type, extension_name):
+        cdef:
+            shared_ptr[CExtensionType] cgen_ext_type
+            c_string c_extension_name
+
+        c_extension_name = tobytes(extension_name)
+        assert storage_type is not None
+        # cgen_ext_type = CGenericExtensionType(
+        #     storage_type.sp_type, c_extension_name)
+        check_status(CGenericExtensionType.FromClass(
+            storage_type.sp_type, c_extension_name, type(self),
+            &cgen_ext_type))
+        self.init(<shared_ptr[CDataType]> cgen_ext_type)
+
+    cdef void init(self, const shared_ptr[CDataType]& type) except *:
+        BaseExtensionType.init(self, type)
+        self.cgen_ext_type = <const CGenericExtensionType*> type.get()
+        # Store weakref on C++ type instance
+        check_status(self.cgen_ext_type.SetInstance(self))
+
+    def __arrow_ext_serialize__(self):
```

Comment:
is there a reason why not to use `def serialize()` and `def deserialize()`? It would be a little nicer for the end user to override

Reply:
I commented above about this. I had thought that serialize/deserialize would be intuitive -- now that these are pseudo-public APIs we might take the opportunity to change. Interested in @pitrou's thoughts